include/wil/cppwinrt.h

```diff
@@ -335,4 +335,210 @@ namespace wil
     }
 }
 
+#if (defined(__cpp_lib_coroutine) && (__cpp_lib_coroutine >= 201902L)) || defined(_RESUMABLE_FUNCTIONS_SUPPORTED)
+
+#if defined(_RESUMABLE_FUNCTIONS_SUPPORTED)
+#include <experimental/coroutine>
+namespace wil::details
+{
+    template<typename T = void> using coroutine_handle = std::experimental::coroutine_handle<T>;
+}
+#else
+#include <coroutine>
+namespace wil::details
+{
+    template<typename T = void> using coroutine_handle = std::coroutine_handle<T>;
+}
+#endif
+
+/// @cond
+
+namespace winrt::Windows::UI::Core
+{
+    struct CoreDispatcher;
+    enum class CoreDispatcherPriority;
+    struct DispatchedHandler;
+}
+
+namespace winrt::Windows::System
+{
+    struct DispatcherQueue;
+    enum class DispatcherQueuePriority;
+    struct DispatcherQueueHandler;
+}
+
+namespace winrt::Microsoft::System
+{
+    struct DispatcherQueue;
+    enum class DispatcherQueuePriority;
+    struct DispatcherQueueHandler;
+}
+
+namespace winrt::Microsoft::UI::Dispatching
+{
+    struct DispatcherQueue;
+    enum class DispatcherQueuePriority;
+    struct DispatcherQueueHandler;
+}
+/// @endcond
+
+namespace wil
+{
+    /// @cond
+    namespace details
+    {
+        struct dispatcher_RunAsync
+        {
+            template<typename Dispatcher, typename... Args>
+            static void Queue(Dispatcher const& dispatcher, Args&&... args)
+            {
+                dispatcher.RunAsync(std::forward<Args>(args)...);
+            }
+        };
+
+        struct dispatcher_TryEnqueue
+        {
+            template<typename Dispatcher,typename... Args>
+            static void Queue(Dispatcher const& dispatcher, Args&&... args)
+            {
+                dispatcher.TryEnqueue(std::forward<Args>(args)...);

```

Comment:
> dispatcher.TryEnqueue(std::forward<Args>(args)...); [](http://example.com/codeflow?start=15&length=52)

Should we be checking the return value here? Or does the wrapper struct convert the bool into an exception somehow?

Reply:
We already designed for the case where `DispatcherQueue.TryEnqueue` and `CoreDispatcher.RunAsync` might accept the work item and then later change their mind and release it without ever invoking it. If `TryEnqueue` returns false, that same "changed my mind" logic will kick in: The `DispatcherQueueHandler` is destructed without ever being invoked, meaning that the work item failed to run, and the coroutine should resume with an exception.

Reply:
Since Jon and I both missed that, worth adding a brief comment explaining why we ignore the return value? This code is tricky!